File: server/internal/store/set_id_vectors_test.go
package store_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/MohammadBnei/wird/server/internal/store"
)

// vectorsPath is the one file both halves answer to. The device suite reads
// the same rows from app/test/data/set_id_vectors_test.dart; neither suite
// computes what it asserts, so the two cannot agree with themselves while
// disagreeing with each other.
const vectorsPath = "../../../docs/adr/0002-set-identity-vectors.json"

type setIDVector struct {
	Label         string `json:"label"`
	ReadingOrder  string `json:"reading_order"`
	StartAyahID   int    `json:"start_ayah_id"`
	EndAyahID     int    `json:"end_ayah_id"`
	ExpectedSetID string `json:"expected_set_id"`
}

func setIDVectors(t *testing.T) []setIDVector {
	t.Helper()
	raw, err := os.ReadFile(vectorsPath)
	if err != nil {
		t.Fatalf("the shared set id vectors are unreadable, so nothing checks the server against the device: %v", err)
	}
	var file struct {
		Vectors []setIDVector `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatalf("%s is not readable as vectors: %v", filepath.Base(vectorsPath), err)
	}
	if len(file.Vectors) == 0 {
		t.Fatal("the shared vectors are empty, so this test passes without checking anything")
	}
	return file.Vectors
}

// The failure: the server names a set by an id no device derives, so every
// prayer a real reader sends is refused — permanently, which dead-letters it —
// and the prayer is lost. The expectations here are checked in, not computed,
// so this test reddens the moment the server's derivation drifts from the one
// the device and the ADR publish.
func TestTheServerDerivesTheSetIdEveryDeviceDerivesForTheSameRange(t *testing.T) {
	for _, v := range setIDVectors(t) {
		if got := store.SetID(v.ReadingOrder, v.StartAyahID, v.EndAyahID); got != v.ExpectedSetID {
			t.Errorf("%s (%s %d-%d): the server calls it %s, every other implementation calls it %s",
				v.Label, v.ReadingOrder, v.StartAyahID, v.EndAyahID, got, v.ExpectedSetID)
		}
	}
}

File: app/test/data/set_id_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:wird/data/sets.dart';

/// The one file both halves answer to. The server suite reads the same rows
/// from server/internal/store/set_id_vectors_test.go; neither suite computes
/// what it asserts, so the two cannot agree with themselves while disagreeing
/// with each other.
const _vectorsPath = '../docs/adr/0002-set-identity-vectors.json';

void main() {
  // The failure: the device names a set by an id the server does not derive,
  // so the server refuses the prayer op — and a refusal is permanent, so the
  // prayer is dead-lettered and the reader loses it for good. The
  // expectations here are checked in, not computed, so this test reddens the
  // moment the device's derivation drifts from the server's.
  test('the device derives the set id the server derives for the same range,'
      ' so the prayer is not refused and lost', () {
    final file = File(_vectorsPath);
    expect(
      file.existsSync(),
      isTrue,
      reason: 'without the shared vectors nothing checks the device against '
          'the server',
    );
    final vectors =
        (jsonDecode(file.readAsStringSync())['vectors'] as List).cast<Map<String, dynamic>>();
    expect(
      vectors,
      isNotEmpty,
      reason: 'empty vectors would pass without checking anything',
    );

    for (final v in vectors) {
      final order = ReadingOrder.values.byName(v['reading_order'] as String);
      expect(
        setIdFor(order, v['start_ayah_id'] as int, v['end_ayah_id'] as int),
        v['expected_set_id'],
        reason: '${v['label']}: the device names this set something the '
            'server will refuse',
      );
    }
  });
}
